extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/DefaultAuthFailureHandler.java

```diff
@@ -0,0 +1,138 @@
+            event.response().end();
+        } else {
+            event.fail(throwable);
+        }
+    }
+
+    /**
+     * Step after {@link HttpAuthenticationMechanism#sendChallenge(RoutingContext)}.
+     *
+     * @param authenticationFailedException {@link AuthenticationFailedException}
+     * @param event {@link RoutingContext}
+     */
+    protected void proceedAfterSendChallenge(AuthenticationFailedException authenticationFailedException,
+            RoutingContext event) {
+        event.end();
+    }
+
+    /**
+     * Skip handling of an event, f.e. when we have already handled the event.
+     *
+     * @return true when handling should be skipped
+     */
+    protected boolean skip(RoutingContext event) {
+        return false;
+    }
+
+    private HttpAuthenticator getAuthenticator(RoutingContext event) {
+        return requireNonNull(
+                event.get(HttpAuthenticator.class.getName()),
+                String.format("'%s' is null. Please add it to the routing context.", HttpAuthenticator.class.getName()));
+    }
+
+    private Throwable extractRootCause(Throwable throwable) {
+        while ((throwable instanceof CompletionException && throwable.getCause() != null) ||
+                (throwable instanceof CompositeException)) {
+            if (throwable instanceof CompositeException) {
+                throwable = ((CompositeException) throwable).getCauses().get(0);
+            } else {
+                throwable = throwable.getCause();
+            }
+        }
+        return throwable;
+    }
+
+    /**
+     * The {@link DefaultAuthFailureHandler} implementation that fail event and let failure handlers down the stream
+     * handle the event.
+     */
+    public static final class FailingDefaultAuthFailureHandler extends DefaultAuthFailureHandler {
+        /**
+         * Signal {@link AuthenticationFailedException} has already been handled here.
+         */
+        public static final String AUTH_FAILED_EX_HANDLED = QuarkusHttpUser.AUTH_FAILURE_HANDLER
+                + ".handled.auth-failed-ex";
+
+        @Override
+        protected boolean skip(RoutingContext event) {
+            // If true, we have already handled this exception and know other failure handlers may want to modify
+            // the response. This situation could happen if the default auth failure handler has been previously
+            // invoked without failing the event (via routingContext.get(AUTH_FAILURE_HANDLER) ...), therefore we skip
+            // the handling as despite its name, the default failure handler is invoked directly (not as event's failure
+            // handler) and calling next would skip actual failure handler.
+            return event.failure() instanceof AuthenticationFailedException
+                    && parseBoolean(event.get(AUTH_FAILED_EX_HANDLED));
+        }
+
+        @Override
+        protected void proceedAfterSendChallenge(AuthenticationFailedException authenticationFailedException,
```

Comment:
I don't like the idea of this handler, after send challenge the request is done. If you want to customize it you need to contribute to the challenge sending process via an auth mechanism, or your need to completely take it over.

Reply:
If you are done after send challenge, it means it won't be possible to catch `AuthenticationFailedException` with failure handler in Reactive Routes. Isn't it reasonable to expect it can be caught? You are going to be able to do that in RESTEasy Reactive and Classic, why should Reactive Routes differ after all? Also I believe users are expecting to be able to customize the response https://github.com/quarkusio/quarkus/issues/5751#issuecomment-633749801

Reply:
If Reactive Routes provided their own handler, it would probably make for cleaner solution than this `proceedAfterSendChallenge`.